File: rust/endo/src/paths.rs
use std::env;
use std::path::PathBuf;

use crate::error::EndoError;

pub struct EndoPaths {
    pub state_path: PathBuf,
    pub ephemeral_path: PathBuf,
    pub sock_path: PathBuf,
    pub cache_path: PathBuf,
    pub log_path: PathBuf,
}

pub fn resolve_paths() -> Result<EndoPaths, EndoError> {
    // If ENDO_STATE_PATH is set, use environment variables for all paths.
    if let Ok(state) = env::var("ENDO_STATE_PATH") {
        let state_path = PathBuf::from(&state);
        let ephemeral_path = env::var("ENDO_EPHEMERAL_STATE_PATH")
            .map(PathBuf::from)
            .unwrap_or_else(|_| state_path.clone());
        let sock_path = env::var("ENDO_SOCK_PATH")
            .map(PathBuf::from)
            .unwrap_or_else(|_| ephemeral_path.join("captp0.sock"));
        let cache_path = env::var("ENDO_CACHE_PATH")
            .map(PathBuf::from)
            .unwrap_or_else(|_| state_path.join("cache"));
        let log_path = state_path.join("endo.log");
        return Ok(EndoPaths {
            state_path,
            ephemeral_path,
            sock_path,
            cache_path,
            log_path,
        });
    }

    // Platform-specific defaults.
    let home = env::var("HOME")
        .map_err(|_| EndoError::Config("HOME not set".to_string()))?;
    let username = env::var("USER")
        .or_else(|_| env::var("LOGNAME"))
        .unwrap_or_else(|_| "unknown".to_string());

    let state_path = if cfg!(target_os = "macos") {
        PathBuf::from(&home).join("Library/Application Support/Endo")
    } else {
        env::var("XDG_STATE_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|_| PathBuf::from(&home).join(".local/state"))
            .join("endo")
    };

    let ephemeral_path = env::var("XDG_RUNTIME_DIR")
        .map(|d| PathBuf::from(d).join("endo"))
        .unwrap_or_else(|_| {
            PathBuf::from(env::temp_dir()).join(format!("endo-{username}"))
        });

    let sock_path = env::var("ENDO_SOCK")
        .map(PathBuf::from)
        .unwrap_or_else(|_| ephemeral_path.join("captp0.sock"));

    let cache_path = if cfg!(target_os = "macos") {
        PathBuf::from(&home).join("Library/Caches/Endo")
    } else {
        env::var("XDG_CACHE_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|_| PathBuf::from(&home).join(".cache"))
            .join("endo")
    };

    let log_path = state_path.join("endo.log");

    Ok(EndoPaths {
        state_path,
        ephemeral_path,
        sock_path,
        cache_path,
        log_path,
    })
}

File: go/engo/daemon/paths.go
package daemon

import (
	"os"
	"os/user"
	"path/filepath"
	"runtime"
)

// EndoPaths holds the resolved filesystem paths for the daemon.
type EndoPaths struct {
	StatePath     string // durable state (applications, capabilities, logs)
	EphemeralPath string // ephemeral state (PID files)
	SockPath      string // Unix domain socket
	CachePath     string // caches
	LogPath       string // daemon log file
}

// ResolvePaths returns platform-appropriate paths for the daemon,
// respecting XDG environment variables. Ported from packages/where/index.js.
//
// When ENDO_STATE_PATH is set, all paths are taken from environment
// variables (ENDO_STATE_PATH, ENDO_EPHEMERAL_STATE_PATH, ENDO_SOCK_PATH,
// ENDO_CACHE_PATH). This allows the test harness to direct engo to
// test-specific temporary directories.
func ResolvePaths() (EndoPaths, error) {
	if statePath := os.Getenv("ENDO_STATE_PATH"); statePath != "" {
		return EndoPaths{
			StatePath:     statePath,
			EphemeralPath: os.Getenv("ENDO_EPHEMERAL_STATE_PATH"),
			SockPath:      os.Getenv("ENDO_SOCK_PATH"),
			CachePath:     os.Getenv("ENDO_CACHE_PATH"),
			LogPath:       filepath.Join(statePath, "endo.log"),
		}, nil
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return EndoPaths{}, err
	}

	u, err := user.Current()
	if err != nil {
		return EndoPaths{}, err
	}
	username := u.Username

	statePath := resolveStatePath(home)
	return EndoPaths{
		StatePath:     statePath,
		EphemeralPath: resolveEphemeralPath(home, username),
		SockPath:      resolveSockPath(home, username),
		CachePath:     resolveCachePath(home),
		LogPath:       filepath.Join(statePath, "endo.log"),
	}, nil
}

func resolveStatePath(home string) string {
	if xdg := os.Getenv("XDG_STATE_HOME"); xdg != "" {
		return filepath.Join(xdg, "endo")
	}
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Application Support", "Endo")
	}
	return filepath.Join(home, ".local", "state", "endo")
}

func resolveEphemeralPath(home, username string) string {
	if xdg := os.Getenv("XDG_RUNTIME_DIR"); xdg != "" {
		return filepath.Join(xdg, "endo")
	}
	if runtime.GOOS == "darwin" {
		tmpdir := os.Getenv("TMPDIR")
		if tmpdir == "" {
			tmpdir = os.TempDir()
		}
		return filepath.Join(tmpdir, "endo-"+username)
	}
	tmpdir := os.Getenv("TMPDIR")
	if tmpdir == "" {
		tmpdir = os.TempDir()
	}
	return filepath.Join(tmpdir, "endo-"+username)
}

func resolveSockPath(home, username string) string {
	if sock := os.Getenv("ENDO_SOCK"); sock != "" {
		return sock
	}
	if xdg := os.Getenv("XDG_RUNTIME_DIR"); xdg != "" {
		return filepath.Join(xdg, "endo", "captp0.sock")
	}
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Application Support", "Endo", "captp0.sock")
	}
	tmpdir := os.Getenv("TMPDIR")
	if tmpdir == "" {
		tmpdir = os.TempDir()
	}
	return filepath.Join(tmpdir, "endo-"+username, "captp0.sock")
}

func resolveCachePath(home string) string {
	if xdg := os.Getenv("XDG_CACHE_HOME"); xdg != "" {
		return filepath.Join(xdg, "endo")
	}
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Caches", "Endo")
	}
	return filepath.Join(home, ".cache", "endo")
}
